### PR TITLE
[SNMP]: Modify test to verify all queue indexes are present

### PR DESCRIPTION
**Pull request by SuvarnaMeenakshi**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305

#### What is the motivation for this PR?
Modify test to detect issue https://github.com/sonic-net/sonic-buildimage/issues/17448.
#### How did you do it?
Currently test_snmp_queue only verifies that each interface does have queue counters associated with it.
It does not verify if all the expected queue counters are present.
Modify test_snmp_queue to add 2 additional checks:
1. Ensure that queue index with queue configuration in config_db, is present for each interface.
2. Ensure that the number of Unicast queue counters for each interface in 'show queue counters <interface name> is the same as the number of queue indexes in the SNMP result.
#### How did you verify/test it?
Test case passes for single-asic/multi-asic/Chassis platforms.

### CODE REVIEW

File: tests/snmp/test_snmp_queue.py [hunk 1]
```diff
@@ -26,25 +28,37 @@ def test_snmp_queues(duthosts, enum_rand_one_per_hwsku_hostname, localhost, cred
...
         # Voq chassis 'QUEUE|<hostname>|<asic_ns>|Ethernet*|2'
         elif len(intf) == 5:
             # Choose only interfaces on current linecard.
             if intf[1] == duthost.hostname:
-                q_interfaces.add(intf[3])
+                intf_idx = 3
+                queue_idx = 4
+        if intf_idx != 0:
+            if intf[intf_idx] in q_interfaces:
+                q_interfaces[intf[intf_idx]].add(intf[queue_idx])
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_queue.py:58]: >add [](http://example.com/codeflow?start=45&length=3)

This line is duplicated in both if/else. To simplify, you just need to initialize the set() if it is empty.

File: tests/snmp/test_snmp_queue.py [hunk 2]
```diff
@@ -62,3 +76,21 @@ def test_snmp_queues(duthosts, enum_rand_one_per_hwsku_hostname, localhost, cred
             intf = alias_port_name_map[v['name']]
             if intf in q_interfaces and 'queues' not in v:
                 pytest.fail("port %s does not have queue counters" % v['name'])
+            # Check if queue index in QUEUE table in config_db
+            # is present in SNMP result
+            if intf in q_interfaces:
+                for queue_idx in q_interfaces[intf]:
+                    snmp_q_idx = int(queue_idx) + 1
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_queue.py:83]: >+ 1 [](http://example.com/codeflow?start=48&length=3)

Explain `+ ` in code comment?

File: tests/snmp/test_snmp_queue.py [hunk 3]
```diff
@@ -62,3 +76,21 @@ def test_snmp_queues(duthosts, enum_rand_one_per_hwsku_hostname, localhost, cred
...
+                    if str(snmp_q_idx) not in v['queues'][direction_type].keys():
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_queue.py:84]: >.keys() [](http://example.com/codeflow?start=73&length=7)

you may remove `.keys()`.

File: tests/snmp/test_snmp_queue.py (hunk 1 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_queue.py]: Updated as suggested

File: tests/snmp/test_snmp_queue.py (hunk 2 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_queue.py:83]: Added comment

File: tests/snmp/test_snmp_queue.py (hunk 3 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_queue.py]: Updated

**Review by qiluo-msft — DISMISSED** [review]

File: tests/snmp/test_snmp_queue.py [hunk 4]
```diff
@@ -62,3 +74,24 @@ def test_snmp_queues(duthosts, enum_rand_one_per_hwsku_hostname, localhost, cred
...
+                    # Increment queue_idx by 1 to form the right OID.
+                    snmp_q_idx = int(queue_idx) + 1
+                    if str(snmp_q_idx) not in v['queues'][direction_type]:
+                        pytest.fail("Expected queue index %d not present in \
+                                     SNMP result for interface %s" % (snmp_q_idx, v['name']))
+            # compare number of unicast queues in CLI to the number of queue indexes in
+            # SNMP result
+            if 'queues' in v:
+                if port_name_to_ns[intf]:
+                    show_cli = 'show queue counters -n {} {} | grep "UC" | wc -l'.format(port_name_to_ns[intf], intf)
```
Comment (qiluo-msft) [inline tests/snmp/test_snmp_queue.py:91]: >show queue counters [](http://example.com/codeflow?start=32&length=19)

There may be a corner case that `show_cli` has counters for this intf, but snmp has no. Then you should fail the testcase.

Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_queue.py:91]: Is it expected that all interfaces in interfaces mib, will have queue counters associated with it, with or without queue configuration?

Comment (qiluo-msft) [reply tests/snmp/test_snmp_queue.py:91]: if `show queue counters` has EthernetN, but snmp has 0 queues on EthernetN, this will pass, but the expectation is failing.

Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_queue.py]: Updated to make sure that all interfaces should have queue counters in SNMP result

**Review by qiluo-msft — APPROVED** [review]
